### PR TITLE
DEV: Add "Particle Swarm Optimization" algorithm to SciPy.optimization

### PR DESCRIPTION
**Pull request by damarro3**
#### What does this implement/fix?
 Added a new optimization algorithm and some tests to ensure
the correct behaviour of the code.

#### Additional information

Also the __init__.py file of
SciPy.optimization has been modified to include this new
feature.

### CODE REVIEW

**Comment by mdhaber** [comment]
@damarro3 thanks for your interest in contributing. I might not be able to review this, but one thing that would help us decide whether to include it would be to compare this against existing global optimizers. Can you also add it to the [benchmarks](http://scipy.github.io/devdocs/dev/contributor/benchmarking.html#benchmarking-with-asv), run them, and share the results?

**Comment by rgommers** [comment]
Hi @damarro3, thanks for opening this PR!

Adding a particle swarm optimization was discussed a few months back on the mailing list, with mildly positive feedback: https://mail.python.org/pipermail/scipy-dev/2020-May/024173.html. Note that there's a detailed reply from @andyfaff to that first message with steps to take for adding a new optimization method. In particular, adding it to the benchmarks and showing that the performance (in terms of number of problems it solves and amount of function evaluations) is competitive.

There was also an unanswered question from me about a comparison to PyGMO.

**Comment by damarro3** [comment]
Thank you @mdhaber and @rgommers, I didn't know about those specific benchmarks. I will implement them and get the results to check if the algorithm is competitive.

Regarding to the PyGMO comparison, I thought about adding this method because is quite used in literature and not everybody install new packages in order to use a general "black box" optimisation method.

**Comment by andyfaff** [comment]
@damarro3, the email topic pointed to by Ralf, is the sequence of steps that we should follow before anything is added. In future, it's best to bring up proposed changes on the scipy-dev mailing list first, before any significant amount of code is written.
For this PR to go any further can you run this code against the global benchmark suite (at least 100 reps) so that we can judge its performance. A solver worth adding will need to have a high percentage of successful minimisations across most of the target problems. Once we know solver performance then code development/review can start

**Comment by andyfaff** [comment]
From my experience adding the `differential_evolution` function the code review process can seem lengthy and very detailed (even nit-picky). However, the large amount of effort it requires to get a new piece of functionality into the package is rewarded in much lower maintenance costs in the long run, and a greater uptake by users.

File: scipy/optimize/_particleswarmoptimization.py [hunk 1]
```diff
@@ -38,151 +37,186 @@ def particle_swarm_optimization(func, bounds, initial_bounds=None, args=None,
         required to have ``len(bounds) == len(x)``. ``len(bounds)`` is used
         to determine the number of parameters in ``x``.
     initial_bounds : sequence or `Bounds`, optional
-        Initial bounds for variables, inside the global search bounds. If it is 
-        not specified, the initial bounds are settle as the global bounds. 
+        Initial bounds for variables, inside the global search bounds. If it is
+        not specified, the initial bounds are settle as the global bounds.
         There are two ways to specify the initial bounds:
         1. Instance of `Bounds` class.
         2. ``(min, max)`` pairs for each element in ``x``, defining the finite
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:44]: I would only accept the `Bounds` class for this, it's a more modern way of doing things.

File: scipy/optimize/_particleswarmoptimization.py [hunk 2]
```diff
@@ -38,151 +37,186 @@ def particle_swarm_optimization(func, bounds, initial_bounds=None, args=None,
...
 
-class PSO:
+    def __init__(self, position, velocity, bestpos, bestof, func):
+        self.position = position
+        self.velocity = velocity
+        self.bestpos = bestpos
+        self.bestof = bestof
+        self.func = func
+
+    def updatevel(self, w, c1, c2, bestgpos, vmin, vmax):
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:199]: update_vel

File: scipy/optimize/_particleswarmoptimization.py [hunk 3]
```diff
@@ -38,151 +37,186 @@ def particle_swarm_optimization(func, bounds, initial_bounds=None, args=None,
...
+        self.velocity = (w * self.velocity +
+                         c1 * (self.bestpos - self.position) +
+                         c2 * (bestgpos - self.position))
+
+        self.velocity = np.maximum(self.velocity, vmin)
+        self.velocity = np.minimum(self.velocity, vmax)
+
+    def updatepos(self, pmin, pmax):
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:207]: update_pos

File: scipy/optimize/_particleswarmoptimization.py [hunk 4]
```diff
@@ -38,151 +37,186 @@ def particle_swarm_optimization(func, bounds, initial_bounds=None, args=None,
...
+        self.position = self.position + self.velocity
+
+        self.position = np.maximum(self.position, pmin)
+        self.position = np.minimum(self.position, pmax)
+
+        pos_value = self.func(self.position)
+        if pos_value <= self.bestof:
+            self.bestof = pos_value
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:215]: best_of

File: scipy/optimize/_particleswarmoptimization.py [hunk 5]
```diff
@@ -38,151 +37,186 @@ def particle_swarm_optimization(func, bounds, initial_bounds=None, args=None,
...
+            self.bestpos = self.position
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:216]: best_pos.
Although if best_pos is actually x, then why not use best_x?

File: scipy/optimize/_particleswarmoptimization.py [hunk 6]
```diff
@@ -307,22 +361,32 @@ def _test(self):
...
+        if self.n_particles <= 0:
             raise ValueError('The number of particles must be greater'
                              ' than 0')
-        if self.num_particles > self.max_iter:
-            raise ValueError('The number of particles must be smaller'
-                             ' than max_iter')
 
         # test w
-        if self.w <= 0:
+        if (((self.velocity_strategy == 'constant') or (self.velocity_strategy == 'random'))
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:373]: self.velocity_strategy in ['constant', 'random']:

File: scipy/optimize/_particleswarmoptimization.py [hunk 7]
```diff
@@ -334,93 +398,103 @@ def _test(self):
...
+    def _particle_initialization(self, i):
+        """ Initialization of particle values """
+        init_position = self._particle_random_generation(self.bounds)
+        self.particles[i] = Particle(position=init_position,
+                                     velocity=0,
+                                     bestpos=init_position,
+                                     bestof=self.func(init_position),
+                                     func=self.func)
+
+    def _update(self):
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:441]: It'd be preferable if the object was iterable, implementing `__iter__` and `__next__`.

File: scipy/optimize/_particleswarmoptimization.py [hunk 8]
```diff
@@ -334,93 +398,103 @@ def _test(self):
...
+        """ Update of the swarm particles """
+        for i in range(self.n_particles):
+            self._particle_update(i)
+            if self.particles[i].bestof <= self.bestgof:
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:445]: Is this parallelisable? If so, then we should use a `workers` keyword to enable parallelisation (see differential_evolution).

File: scipy/optimize/_particleswarmoptimization.py [hunk 9]
```diff
@@ -334,93 +398,103 @@ def _test(self):
...
+                self.bestgpos = self.particles[i].bestpos
+                self.bestgof = self.particles[i].bestof
+
+        self.fevals += self.n_particles
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:449]: rename variable to `nfev`.

File: scipy/optimize/_particleswarmoptimization.py [hunk 10]
```diff
@@ -334,93 +398,103 @@ def _test(self):
...
+        self.it += 1
```
Comment (andyfaff) [inline scipy/optimize/_particleswarmoptimization.py:450]: rename attribute to `nit`

**Review by andyfaff — COMMENTED** [review]
I just glanced at a few things, so it's not a comprehensive review. Before you go any further can you set it up to [run against the global optimization benchmarks](https://github.com/scipy/scipy/blob/master/benchmarks/benchmarks/optimize.py#L211)? New global optimizers need to be run against the benchmark suite to show that they offer a performance advantage to the existing minimizer functions in order for them to be merged.
i.e. a greater rate of success, or smaller amount of function evaluations (for a reasonable success), etc.

**Comment by damarro3** [comment]
I have executed the benchmarks. As soon as they finish, I will post the results.

**Pull request closed without merging**